zkp-component/src/zkp_prover.rs

```diff
@@ -165,45 +134,102 @@ impl<N: Network> ZKProver<N> {
-            match proof {
-                Ok((new_zkp_state, block)) => {
-                    assert!(
-                        new_zkp_state.latest_proof.is_some(),
-                        "The generate new proof should never produces a empty proof"
-                    );
-                    let zkp_state_lock = self.zkp_state.upgradable_read();
-
-                    // If we received a more recent proof in the meanwhile, we should have cancelled the proof generation process already.
-                    assert!(
-                        zkp_state_lock.latest_block_number < new_zkp_state.latest_block_number,
-                        "The generated proof should always be more recent than the current state"
-                    );
-
-                    let mut zkp_state_lock = RwLockUpgradableReadGuard::upgrade(zkp_state_lock);
-                    *zkp_state_lock = new_zkp_state;
-
-                    let zkp_state_lock = RwLockWriteGuard::downgrade(zkp_state_lock);
-
-                    let proof: ZKProof = zkp_state_lock.clone().into();
-                    Self::broadcast_zk_proof(&self.network, proof.clone());
-                    return Poll::Ready(Some((proof, block)));
-                }
-                Err(e) => {
-                    log::error!("Error generating ZK Proof for block {}", e);
-                }
-            };
+        if let Some(ref mut proof_future) = self.proof_future {
+            if let Poll::Ready(proof) = proof_future.poll_unpin(cx) {
+                self.proof_future = None;
+                self.sender = None;
+                match proof {
+                    Ok((new_zkp_state, block)) => {
+                        assert!(
+                            new_zkp_state.latest_proof.is_some(),
+                            "The generate new proof should never produces a empty proof"
+                        );
+                        let zkp_state_lock = self.zkp_state.upgradable_read();
+
+                        // If we received a more recent proof in the meanwhile, we should have cancelled the proof generation process already.
+                        assert!(
+                            zkp_state_lock.latest_block_number < new_zkp_state.latest_block_number,
+                            "The generated proof should always be more recent than the current state"
+                        );
+
+                        let mut zkp_state_lock = RwLockUpgradableReadGuard::upgrade(zkp_state_lock);
+                        *zkp_state_lock = new_zkp_state;
+
+                        let zkp_state_lock = RwLockWriteGuard::downgrade(zkp_state_lock);
+
+                        let proof: ZKProof = zkp_state_lock.clone().into();
+                        Self::broadcast_zk_proof(&self.network, proof.clone());
+                        return Poll::Ready(Some((proof, block)));
+                    }
+                    Err(e) => {
+                        log::error!("Error generating ZK Proof for block {}", e);
```

Comment:
I'd use tracing style for this
```suggestion
                        log::error!(error = e, "Error generating ZK Proof for block");
```